proxy/http/remap/NextHopConsistentHash.cc

```diff
@@ -214,244 +234,230 @@ NextHopConsistentHash::findNextHop(TSHttpTxn txnp, void *ih, time_t now)
 {
   uint32_t const NO_RING_USE_POST_REMAP = uint32_t(0) - 1;
 
-  HttpSM *sm                   = reinterpret_cast<HttpSM *>(txnp);
-  ParentResult *result         = &sm->t_state.parent_result;
-  HttpRequestData request_info = sm->t_state.request_data;
-  int64_t sm_id                = sm->sm_id;
-  int64_t retry_time           = sm->t_state.txn_conf->parent_retry_time;
-  time_t _now                  = now;
-  bool firstcall               = false;
-  bool nextHopRetry            = false;
-  bool wrapped                 = false;
+  HttpSM *sm                    = reinterpret_cast<HttpSM *>(txnp);
+  ParentResult &result          = sm->t_state.parent_result;
+  HttpRequestData &request_info = sm->t_state.request_data;
+  int64_t sm_id                 = sm->sm_id;
+  int64_t retry_time            = sm->t_state.txn_conf->parent_retry_time;
+  time_t _now                   = now;
+  bool firstcall                = false;
+  bool nextHopRetry             = false;
+  bool wrapped                  = false;
   std::vector<bool> wrap_around(groups, false);
-  uint32_t cur_ring = 0; // there is a hash ring for each host group
-  uint64_t hash_key = 0;
-  uint32_t lookups  = 0;
-  ATSHash64Sip24 hash;
-  HostRecord *hostRec              = nullptr;
+  uint32_t cur_ring                = 0; // there is a hash ring for each host group
+  uint32_t lookups                 = 0;
   std::shared_ptr<HostRecord> pRec = nullptr;
   HostStatus &pStatus              = HostStatus::instance();
   TSHostStatus host_stat           = TSHostStatus::TS_HOST_STATUS_INIT;
   HostStatRec *hst                 = nullptr;
   Machine *machine                 = Machine::instance();
+  std::string_view first_call_host;
+  int first_call_port = 0;
 
-  if (result->line_number == -1 && result->result == PARENT_UNDEFINED) {
+  if (result.line_number == -1 && result.result == PARENT_UNDEFINED) {
     firstcall = true;
   }
 
+  // firstcall indicates that this is the first time the state machine has called findNextHop() for this
+  // particular transaction so, a parent will be looked up using a hash from the request to locate a
+  // parent on the consistent hash ring.  If not first call, then the transaction was unable to use the parent
+  // returned from the "firstcall" due to some error so, subsequent calls will not search using a hash but,
+  // will instead just increment the hash table iterator to find the next parent on the ring
   if (firstcall) {
-    NH_Debug(NH_DEBUG_TAG, "[%" PRIu64 "] firstcall, line_number: %d, result: %s", sm_id, result->line_number,
-             ParentResultStr[result->result]);
-    result->line_number = distance;
-    cur_ring            = 0;
+    NH_Debug(NH_DEBUG_TAG, "[%" PRIu64 "] firstcall, line_number: %d, result: %s", sm_id, result.line_number,
+             ParentResultStr[result.result]);
```

Comment:
Because of this change we have to change the 31 lines line this one:  https://github.com/apache/trafficserver/blob/44d1be681245f73e82eee59699ad0660dea91290/tests/gold_tests/next_hop/zzz_strategies_peer2/trace.gold#L3 in the gold file from result->result to result.result .

Reply:
Also have to change the 48 line like this one:  https://github.com/apache/trafficserver/blob/44d1be681245f73e82eee59699ad0660dea91290/tests/gold_tests/next_hop/zzz_strategies_peer/trace.gold#L3

Reply:
@ywkaras I changed the text of the Debug at line 448 back so that it logs 'result->result'